### PR TITLE
crimson/os/seastore: lba pointer consolidation

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/lba_manager/btree/lba_btree_node.h [hunk 1]
```diff
@@ -171,6 +171,8 @@ struct LBALeafNode
 
     for (auto i : *this) {
       auto child = (LogicalCachedExtent*)this->children[i.get_offset()];
+      // Children may not be marked as stable yet,
+      // the specific order is undefined in the transaction prepare record phase.
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/lba_btree_node.h:175]: Is it also good to rename validate_stable_children() to validate_(stable_)children_upon_preparing() ?

File: src/crimson/os/seastore/btree/fixed_kv_btree.h [hunk 2]
```diff
@@ -588,6 +588,13 @@ class FixedKVBtree {
                 : true);
             }
           }
+        } else if (child == RESERVATION_PTR) {
+          if constexpr(
+            !std::is_base_of_v<typename internal_node_t::base_t,
+                               child_node_t>) {
+            assert(i->get_val().pladdr.is_paddr());
+            assert(i->get_val().pladdr.get_paddr() == P_ADDR_ZERO);
+          }
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_btree.h:597]: IIUC the `else` case means that `node` is leaf, and I suppose a similar check to P_ADDR_ZERO can also be applied there ?

Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_btree.h:597]: Is `check_node()` supposed to be called any time, including trans pending and committing time?

**Review by cyx1231st — COMMENTED** [review]
I'll try to summarize from https://github.com/ceph/ceph/pull/47749 about what comment might still be valid and worth doing, and move the "crimson-has-followup" label here to reflect our latest consensus.

**Comment by cyx1231st** [comment]
> I'll try to summarize from https://github.com/ceph/ceph/pull/47749 about what comment might still be valid and worth doing, and move the "crimson-has-followup" label here to reflect our latest consensus.

I'm working on some other simple cleanups based on the previous comments, will summarize after that.

File: src/crimson/os/seastore/cached_extent.h [hunk 3]
```diff
@@ -322,6 +322,24 @@ class CachedExtent
     return true;
   }
 
+  void rewrite(CachedExtent &e, extent_len_t o) {
+    if (!e.is_pending()) {
+      prior_instance = &e;
+    } else {
+      assert(e.is_mutation_pending());
+      prior_instance = e.get_prior_instance();
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:312]: Ah, will correct it.

**Comment by cyx1231st** [comment]
@xxhdx1985126 

Regarding to https://github.com/ceph/ceph/blob/639d182732644edc5c413562ebc904ab5b953303/src/crimson/os/seastore/btree/fixed_kv_node.h#L52-L83

The comment seems insufficient and I'm not sure about how the current implementation is supposed to handle the other cases, for example:
* How is exist_clean is supposed to be handled?
* How is exist_mutation_pending is supposed to be handled?
* In case of inplace-rewrite @myoungwon, seems to me it is not the exact case to `3. if nodes are initial_pending` because its parent won't become pending.
* I suppose exist_clean can also be special if its parent doesn't need to become pending.

I understand that these features were introduced after the lba-pointer PR, but it doesn't seem that we have discussed about them very carefully. They are passing our unit tests, but the tests may be insufficient in case of transaction conflicts.

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126
> 
> Regarding to
> 
> https://github.com/ceph/ceph/blob/639d182732644edc5c413562ebc904ab5b953303/src/crimson/os/seastore/btree/fixed_kv_node.h#L52-L83
> 
> The comment seems insufficient and I'm not sure about how the current implementation is supposed to handle the other cases, for example:
> 
> * How is exist_clean is supposed to be handled?
> * How is exist_mutation_pending is supposed to be handled?
> * In case of inplace-rewrite @myoungwon, seems to me it is not the exact case to `3. if nodes are initial_pending` because its parent won't become pending.
> * I suppose exist_clean can also be special if its parent doesn't need to become pending.
> 

IIUC, `EXIST_CLEAN` extents are created by extent remappings, so they must corresponds to lba mapping changes, which means their parents will be pending. `EXIST_MUTATION_PENDING` extents come from `EXIST_CLEAN` extents, so their parents are also pending. Am I missing something?

On the other hand, it seems that inplace-rewrite doesn't actually create new extents like other rewriting process, so the pointer links between them and their parents don't need to be touched, which means it's ok no to make their parents pending or modified the pointers in their parents. Am I right?

https://github.com/ceph/ceph/blob/639d182732644edc5c413562ebc904ab5b953303/src/crimson/os/seastore/transaction_manager.cc#L557-L561

> I understand that these features were introduced after the lba-pointer PR, but it doesn't seem that we have discussed about them very carefully. They are passing our unit tests, but the tests may be insufficient in case of transaction conflicts.

**Comment by cyx1231st** [comment]
> IIUC, EXIST_CLEAN extents are created by extent remappings, so they must corresponds to lba mapping changes, which means their parents will be pending.

So it should be correct to replace `initial_pending` by `initial_pending or exist_clean` in the above comment?

If that's correct, I'm still wondering why there are many related logic only under the condition `is_initial_pending()`, instead of both `is_initial_pending() || is_exist_clean()` ?

**Comment by cyx1231st** [comment]
> On the other hand, it seems that inplace-rewrite doesn't actually create new extents like other rewriting process, so the pointer links between them and their parents don't need to be touched, which means it's ok no to make their parents pending or modify the pointers in their parents. Am I right?

Reasonable to me. Hm, probably good to mention this in the comment.

**Comment by cyx1231st** [comment]
> I suppose exist_clean can also be special if its parent doesn't need to become pending.

Ah, it's not true, because the length in the parent entry must be changed at the same time.

(Unrelated, but probably it worths to ensure that we don't remap an extent to another extent with the same length)

**Comment by xxhdx1985126** [comment]
> > IIUC, EXIST_CLEAN extents are created by extent remappings, so they must corresponds to lba mapping changes, which means their parents will be pending.
> 
> So it should be correct to replace `initial_pending` by `initial_pending or exist_clean` in the above comment?
> 
> If that's correct, I'm still wondering why there are many related logic only under the condition `is_initial_pending()`, instead of both `is_initial_pending() || is_exist_clean()` ?

Just scanned the code, it seems that almost all of the checks against `is_initial_pending()` happen on non-object-data extents, the rest happen in cases where `EXIST_CLEAN` extents don't need to be considered:

```
src/crimson/os/seastore/transaction.h:    } else if (ref->is_initial_pending()) {
src/crimson/os/seastore/btree/fixed_kv_node.h:    ceph_assert(dest.is_initial_pending());
src/crimson/os/seastore/btree/fixed_kv_node.h:      ceph_assert(src.is_initial_pending());
src/crimson/os/seastore/btree/fixed_kv_node.h:    return is_initial_pending() && get_prior_instance();
src/crimson/os/seastore/btree/fixed_kv_node.h:    if (this->is_initial_pending()) {
src/crimson/os/seastore/btree/fixed_kv_node.h:   * physical node is is_initial_pending() or just is_mutable().
src/crimson/os/seastore/btree/fixed_kv_node.h:    if (this->is_initial_pending()) {
src/crimson/os/seastore/btree/fixed_kv_node.h:    if (this->is_initial_pending()) {
src/crimson/os/seastore/btree/fixed_kv_node.h:      if (this->is_initial_pending()) {
src/crimson/os/seastore/btree/fixed_kv_node.h:    assert(this->is_initial_pending()
src/crimson/os/seastore/extent_placement_manager.cc:      if (ex->is_initial_pending()) {
src/crimson/os/seastore/cached_extent.h:  bool is_initial_pending() const {
src/crimson/os/seastore/cached_extent.h:   * Returns current address of extent.  If is_initial_pending(), address will
src/crimson/os/seastore/cached_extent.h:  /// address of original block -- record relative iff is_initial_pending()
src/crimson/os/seastore/cached_extent.h:    if (is_initial_pending() && addr.is_record_relative()) {
src/crimson/os/seastore/backref/btree_backref_manager.cc:    ceph_assert(backref_root->is_initial_pending()
src/crimson/os/seastore/onode_manager/staged-fltree/node_layout.h:      assert(extent->is_initial_pending());
src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h:      if (!extent->is_initial_pending()) {
src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:    if (extent->is_initial_pending()) {
src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:      assert(extent->is_initial_pending());
src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:    assert(!extent->is_initial_pending());
src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h:      assert(fresh_extent->is_initial_pending());
src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:    ceph_assert(lba_root->is_initial_pending()
src/crimson/os/seastore/lba_manager/btree/lba_btree_node.h:    if (this->is_initial_pending()) {
src/crimson/os/seastore/lba_manager/btree/lba_btree_node.h:    if (this->is_initial_pending()) {
src/crimson/os/seastore/transaction_manager.cc:  assert(extent->is_valid() && !extent->is_initial_pending());
```

**Comment by xxhdx1985126** [comment]
> > EXIST_MUTATION_PENDING extents come from EXIST_CLEAN extents, so their parents are also pending.
> 
> So does it mean `EXIST_MUTATION_PENDING` should also belong to the category 3 (initial_pending) as well ? 

For lba pointers yes, but I think it would be better to distinguish it in the comment, will change it.

> Or is it actually better to discuss it separately in the comment ?

**Comment by cyx1231st** [comment]
```
* 2. if nodes are mutation_pending: 
*   a. parent is empty and needs to be fixed upon commit 
```
Is it possible that the parent is initial-pending, and the child is mutation-pending ? (I don't see how it's impossible.)

If that's possible, how does the bi-directional parent-child lookup works if there is no parent-child link in this case ?

Or does this comment actually inaccurate ?

**Comment by xxhdx1985126** [comment]
> ```
> * 2. if nodes are mutation_pending: 
> *   a. parent is empty and needs to be fixed upon commit 
> ```
> 
> Is it possible that the parent is initial-pending, and the child is mutation-pending ? (I don't see how it's impossible.)

Yes, it's possible.

> 
> If that's possible, how does the bi-directional parent-child lookup works if there is no parent-child link in this case ?

IIRC, at present, the lookup of lba/backref tree is a top-down process and doesn't involve the parent link. On the other hand, the maintenance of the corrent link between initial_pending parent and mutation_pending children works as follows:
1. `mutation_pending` extents always take their prior_instance's parent on committing;
2. `initial_pending` lba/backref extents always adjust their stable sources' children's parent links;
3. if the `mutation_pending` extents are committed first, their parent pointer will point to their prior_instances' parents after committing; and the `initial_pending` parents will adjust those once-mutation-pending children's parent pointers so that they point to the new parents;
4. if the `initial_pending` parents are committed first, they will adjust their stable children's parent pointers; and then when the `mutation_pending` children take their prior_instances' parent pointers, they will see parent pointers pointing to the new parents.

In the current `Cache::prepare_record` implemention, I think `mutation_pending` extents are always committed first, so `3` should always be the case.

> 
> Or does this comment actually inaccurate ?

**Comment by cyx1231st** [comment]
>> ```
>> * 2. if nodes are mutation_pending: 
>> *   a. parent is empty and needs to be fixed upon commit 
>> ```
>> Is it possible that the parent is initial-pending, and the child is mutation-pending ? (I don't see how it's impossible.)

> Yes, it's possible.

Sorry my question above might be confusing. From the standup, if the parent is initial-pending and the child is mutation-pending, their parent/child pointers should be empty, so the referred comment is accurate.

File: src/crimson/os/seastore/cached_extent.h [hunk 4]
```diff
@@ -1213,6 +1241,17 @@ class LogicalCachedExtent : public ChildableCachedExtent {
...
 
+  void on_rewrite(CachedExtent &extent, extent_len_t off) final {
+    assert(get_type() == extent.get_type());
+    auto &lextent = (LogicalCachedExtent&)extent;
+    lextent.get_bptr().copy_out(
+      off,
+      get_length(),
+      get_bptr().c_str());
+    set_laddr(lextent.get_laddr() + off);
+    set_modify_time(lextent.get_modify_time());
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1235]: Um, only logical extents have laddr.

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/cached_extent.h:1235]: Yeah, I meant "except" set_laddr above.

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1235]: ah, yes

**Comment by cyx1231st** [comment]
> I'll try to summarize from https://github.com/ceph/ceph/pull/47749 about what comment might still be valid and worth doing, and move the "crimson-has-followup" label here to reflect our latest consensus.

Summarized in https://github.com/ceph/ceph/pull/47749#issuecomment-2003270273 to remember.

**Comment by cyx1231st** [comment]
Not sure if the latest change is ready refering to the unaddressed comments, let me know if needs review.

File: src/crimson/os/seastore/btree/fixed_kv_btree.h (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:601]: Yes

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:601]: Um, it seems that there should be no `else` case, will try to assert it.

**Comment by xxhdx1985126** [comment]
@cyx1231st I think this pr is good to take another round of review, please take a look again, thanks:-)

**Comment by xxhdx1985126** [comment]
jenkins test api

File: src/crimson/os/seastore/btree/fixed_kv_btree.h [hunk 5]
```diff
@@ -592,6 +592,15 @@ class FixedKVBtree {
...
+            !std::is_base_of_v<typename internal_node_t::base_t,
+                               child_node_t>) {
+            assert(i->get_val().pladdr.is_paddr());
+            assert(i->get_val().pladdr.get_paddr() == P_ADDR_ZERO);
+          } else {
+            ceph_abort();
+          }
         }
       } else {
         ceph_abort("impossible");
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_btree.h:606]: May be related questions:

Why `FixedKV[Internal/Leaf]Node::child_pos_for_key()` were implemented differently?

Why `check_node()` doesn't reuse `child_pos_for_key()`?

**Review by cyx1231st — CHANGES_REQUESTED** [review]
Let's take the chance to also amend the comments in `fixed_kv_node.h`:
* 1.c rename to 1.d
* 2.c rename to 2.d

And if I'm correct:
```
4. EXIST_CLEAN and EXIST_MUTATION_PENDING belong to 3 above
    (except that they cannot be rewritten) 
   because their parents must be mutated upon remapping.
```

File: src/crimson/os/seastore/btree/fixed_kv_btree.h (hunk 5 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:615]: > May be related questions:
> 
> Why `FixedKV[Internal/Leaf]Node::child_pos_for_key()` were implemented differently?
> 

I think I was modeling what `FixedKVBtree::lower_bound()` was doing. Seems that this is not necessary, will try to change it.

> Why `check_node()` doesn't reuse `child_pos_for_key()`?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:616]: just fixed.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:616]: All done.

**Comment by xxhdx1985126** [comment]
jenkins test api

File: src/crimson/os/seastore/btree/fixed_kv_btree.h [hunk 6]
```diff
@@ -592,6 +592,15 @@ class FixedKVBtree {
...
           }
+        } else if (child == get_reserved_ptr()) {
+          if constexpr(
+            !std::is_base_of_v<typename internal_node_t::base_t,
+                               child_node_t>) {
+            assert(i->get_val().pladdr.is_paddr());
+            assert(i->get_val().pladdr.get_paddr() == P_ADDR_ZERO);
+          } else {
+            ceph_abort();
+          }
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/btree/fixed_kv_btree.h:608]: After the change, line 572-576 and 579-583 can be removed.

**Review by cyx1231st — APPROVED** [review]
LGTM except 1 comment above.

**Comment by xxhdx1985126** [comment]
> LGTM except 1 comment above.

Done:-)

**Comment by xxhdx1985126** [comment]
>/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/options/legacy_config_opts.h:1:10: fatal error: 'global_legacy_options.h' file not found
#include "global_legacy_options.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test windows

**Comment by cyx1231st** [comment]
jenkins test windows

**Review by cyx1231st — APPROVED** [review]

**Comment by cyx1231st** [comment]
https://shaman.ceph.com/builds/ceph/ci-yingxin-pr-55831/

**Comment by tchaikov** [comment]
> > /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/options/legacy_config_opts.h:1:10: fatal error: 'global_legacy_options.h' file not found
> > #include "global_legacy_options.h"
> > ^~~~~~~~~~~~~~~~~~~~~~~~~
> > 1 error generated.

@xxhdx1985126 if we spot a similar case, could you please paste the link to the jenkins job, and at me? i'd like to fix it.

**Comment by xxhdx1985126** [comment]
> > > /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/options/legacy_config_opts.h:1:10: fatal error: 'global_legacy_options.h' file not found
> > > #include "global_legacy_options.h"
> > > ^~~~~~~~~~~~~~~~~~~~~~~~~
> > > 1 error generated.
> 
> @xxhdx1985126 if we spot a similar case, could you please paste the link to the jenkins job, and at me? i'd like to fix it.

Sorry, I 

> > > /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/options/legacy_config_opts.h:1:10: fatal error: 'global_legacy_options.h' file not found
> > > #include "global_legacy_options.h"
> > > ^~~~~~~~~~~~~~~~~~~~~~~~~
> > > 1 error generated.
> 
> @xxhdx1985126 if we spot a similar case, could you please paste the link to the jenkins job, and at me? i'd like to fix it.

@tchaikov Hi, here's the jenkins job: https://jenkins.ceph.com/job/ceph-pull-requests/135238/consoleFull

**Pull request merged**